### ISSUE TITLE
Update to systemd 245-2 breaks user login for usernames starting with a number

### ISSUE THREAD

**Issue by 3ndymion** [systemd/systemd#15141]
**systemd version the issue has been seen with**
> v245-2 & v245-3

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > Should be able to log in with the display manager.

**Unexpected behaviour you saw**
 > Can no longer log in with the display manager.

**Steps to reproduce the problem**
Update to systemd v245-2 or v245-3.

Hello all.  I have an issue to report, & I see others having a similar issue as well.  I run 2 Arch Linux installations, 1 KDE, & 1 XFCE.  After updating each one to systemd v245-2, they both broke in the same way.  I am no longer able to log in to the system using the display manager.  SDDM throws up an X message saying "Could not sync environment to dbus", & LightDM says "Unable to contact settings server Could not connect: Connection refused".

I have never experienced these problems with systemd 244 & below.  I don't know what changed with these new versions, but it definitely broke both of my Arch installations.  I have the journalctl -b boot logs from each desktop environment.  I will attach them as files since they are so long.

[loginfail_KDE_EDITED.log](https://github.com/systemd/systemd/files/4340980/loginfail_KDE_EDITED.log)
[loginfail_XFCE_EDITED.log](https://github.com/systemd/systemd/files/4340981/loginfail_XFCE_EDITED.log)

**Comment by loqs** [comment]
Downstream bug report https://bugs.archlinux.org/task/65825

Issue appears limited to invalid login names such as 3ndymion

**Comment by 3ndymion** [comment]
@loqs Thank you for linking, I ran out of time now.

Yes, please see the report I filed over there.  In my case, a new "normal" user name account was allowed to log in with the display manager.  If the intention is to not allow a user with a "bad name" to log in, I think the idea is a bit flawed & should be rethought.  If a user already exists, they should be allowed to log in regardless.  Too much to type again, please see the link.

To all who see this, thank you even just for looking.

**Comment by mbiebl** [comment]
```
$ adduser 3dtest
...
<then try to log in as that user>

Mär 17 03:08:07 pluto gdm-password][6607]: pam_systemd(gdm-password:session): Failed to get user record: Das Argument ist ungültig
```

**Comment by floppym** [comment]
https://github.com/systemd/systemd/blob/v245/src/basic/user-util.c#L710

**Comment by floppym** [comment]
The upstream version of `useradd` does not allow usernames to start with a digit. Debian patches it to allow this.

https://salsa.debian.org/debian/shadow/-/blob/master/debian/patches/506_relaxed_usernames

**Comment by floppym** [comment]
The validation code in systemd has already been relaxed once to allow periods in user/group names in certain contexts. Perhaps it could be further relaxed.

https://github.com/systemd/systemd/commit/1a29610f5fa1bcb2eeb37d2c6b79d8d1a6dbb865

**Comment by 3ndymion** [comment]
Thank you for all that knowledge.  It's very insightful.  Since this is narrowed down to the single issue of having a number as the first character, I think it would be very helpful to have that restriction relaxed as well.  I do not yet know C or C++, but I don't see any harm that could come from a user name like this.  It does not pose any type of threat for shell operations with variable expansion of any kind.

In all the years I've been using different Linux distros, I can only remember 1 time when I had trouble with my user name.  I do not remember if it was with systemd or in OpenWRT, but Transmission did not want to start running because of my user name.  In that case, I was able to tell it to use my UID number instead of my name, & it worked fine.

Normally, I strive to put out works that will be helpful to others.  I feel really bad pushing for this when I seem to be the only one hit by it, but to expand the thinking a bit, this could actually be quite useful for any number of people.

### Some Reasoning
Using myself as an example, I have my goals set on both programming & public media entertainment.  Concentrating on the entertainment part, it is often a good idea for anyone with that goal in mind to try to protect themselves a bit.  Using a fake name or screen name as their online persona is a good way to do that.  But what is the need to protect oneself???

Well, it can be just for the sake of privacy.  Some people seek to be famous, but others do not.  Using a screen name as their online persona, instead of one's real name, can help to protect one's personal privacy, even if just a bit.

However, more importantly, there is also the issue of safety.  When one becomes a public figure, whether big or small, it is very easy to fall into some amount of danger.  It only takes 1 fan or fellow friend to bring hell to your front door in their raging jealousy.  From online bullying to doxxing to real life danger, anything can happen.  I have seen such things happen to many people in the entertainment field, & it can be quite scary.

Now, to tie this into our world of computing, it is not too rare for people to have screen names that begin with a number, like my own.  When we create many digital works, it is much better & easier to have everything organized in our own home that is named the same as our online persona.  Having our user account retain the same name as our online persona is highly beneficial, both to help avoid confusion, & to avoid any other types of headaches that may arise from having to use names with different spellings.

Actually, there is also a real name example I can come up with.  If someone by the name of Zakhar wants to use Linux, he will probably be disappointed to find that he cannot use his native name, Захар.  There are probably other similar names that I do not know about as well.

https://en.wikipedia.org/wiki/Zakhar

### In Summary
Considering the fact that such a name would not cause any real world damage (to my knowledge), I think that a relaxing of this restriction would be beneficial to anyone who may need it.  Please do consider it.  I think it would be helpful for people in the future, & I can say for a fact that it certainly would be helpful for my own self right now.   :innocent:

**Comment by mbiebl** [comment]
I think it makes a lot of sense to be picky and restrictive when creating users but when reading existing users the rules should be much more relaxed.

**Comment by eli-schwartz** [comment]
> The upstream version of `useradd` does not allow usernames to start with a digit. Debian patches it to allow this.

This is invalid, incorrect information.

The upstream shadow project has a --badnames argument which provides a publicly sanctioned interface for creating such usernames on *any* distribution, Debian or otherwise.

From the Debian patch:

       On Debian, the only constraints are that groupnames must neither start
       with a dash ('-') nor plus ('+') nor tilde ('~') nor contain a
       colon (':'), a comma (','), or a whitespace (space:' ',
       end of line: '\n', tabulation: '\t', etc.).

The --badnames option will allow you to create a user that violates even this relaxed rule. Once it is in the passwd database, systemd must not deny it.

**Comment by 3ndymion** [comment]
I got the update from the Arch Linux repos, & all is well again.  Thank you all for working on this.  It's a very, very big help to me, & it's very much appreciated.  Thank you so much.  :smiley:

**Comment by cheretbe** [comment]
Arrived here after joining an Ubuntu 20.04 machine (systemd 245.4) to a domain, that contains usernames not only starting with digits, but **consisting entirely from digits**, with no other symbols at all (like `32421` etc.). Is there any possibility to use such names?

**Comment by keszybz** [comment]
> Arrived here after joining an Ubuntu 20.04 machine (systemd 245.4) to a domain, that contains usernames not only starting with digits, but **consisting entirely from digits**, with no other symbols at all (like `32421` etc.). Is there any possibility to be able to login with such names?

Names that are indistinguishable from numbers are not going to be supported by systemd, sorry.

**Comment by DwarfFighterCleric** [comment]
Can confirm that this is still happening even in Live-Sessions, booting latest updated Nov-2020 builds of Manjaro and MX ISO's.  
I was trying to pin point what exactly is preventing me from migrate and apply different desktop settings across multiple machines...  
more detailed here: https://forum.xfce.org/viewtopic.php?pid=60770#p60770  
...and there was even an occasion where simply creating a new user and logging in with it would not work.  
I was stuck in this system using the Live-Session user, couldn't create a new user because X server would not work anymore.  
Has anyone come across desktop settings not being saved across reboots? Any workarounds for it?  
I'm asking here because you guys seem to have been dealing with this for a long time, maybe faster to ask here than browse hundreds of google result pages.

**Comment by romain-marie** [comment]
> > Arrived here after joining an Ubuntu 20.04 machine (systemd 245.4) to a domain, that contains usernames not only starting with digits, but **consisting entirely from digits**, with no other symbols at all (like `32421` etc.). Is there any possibility to be able to login with such names?
> 
> Names that are indistinguishable from numbers are not going to be supported by systemd, sorry.

Why ?

**Comment by poettering** [comment]
> Why ?

Isn't that obvious? Because many tools parse user specifications as both numeric UID and resolve it as user name, and it's ambiguous whether a specification like "0" (or "1000") then would mean the user with the numeric UID 0 (or 1000), or whether it references a user by the name "0" (or "1000") that might have the numeric UID 5555...

Take "chmod" for example: it accepts both numeric user specifications and those by their user name. And tools that implement that sometimes resolve as name first, and sometimes parse as numeric UID first, and this you have a lot of ambiguity if you allow that. Hence we refuse it.

**Comment by romain-marie** [comment]
Thank you for your reply and explanation. I understand the difficulty to handle properly logins containing only digits, because they may be understood as uid. But I think a better option would be to choose a strategy like chmod or other commands, that try to resolve name or uid first. 
The main problem for us : it was working properly in the past with systemd. 
It's a pity to remove support of this, because now user can authenticate properly, but desktop is broken because systemd refuses to handle username.

**Comment by cheretbe** [comment]
Actually I've been joining Ubuntu systems to this particular domain (that uses digit-only user names) without any issue for a long time. Desktop versions 16.04 and 18.04 work just fine, as well as server versions 16.04, 18.04 and 20.04. I just used relative high numbers (>10000) for `idmap` parameter in `smb.conf`. Apparently chmod and other utilities are smart enough to try a parameter as an UID and then fall back to using it as a user name. At least I've never had a problem with this. Not a single time.
But systemd has it's own way in every aspect of Linux life, go figure...

**Comment by mindrocker** [comment]
At our University numerical login IDs have been the norm for many years. 

The dominant platform is Windows (and where the numerical ID requirements is enforced); but, also we have a substantial Linux infrastructure - and allot of students & academics that depend on it.

Our current deployment (Debian Buster / Gnome 3) is running systemd 241, and there is no issues with logins. 

Next release was to be Debian Bullseye. But with systemd 247 and the lack of numerical ID support, we've been a bit broadsided by this decision.

The dropping of support for numerical IDs will massively affect us, and our ability to offer Linux as an open source alternative to Windows.

Is there absolutely no chance this could be reassessed?

**Comment by empereira** [comment]
Like @mindrocker, in our institution, we also use Linux distributions for our students. Thus, offering opensource software so that the community can have contact with this type of distribution.

At the moment I am updating ubuntu 18.0x to 20.04, which has this obstacle.

Is there any way to apply some patch to use version 20.04 accepting logins only numbers?

**Comment by keszybz** [comment]
You can revert those patches locally. This is open source after all.

**Comment by empereira** [comment]
> You can revert those patches locally. This is open source after all.

Which files should be patched?
